File: branches/Medusa/Source/Fantasy.Jobs/Management/StandaloneJobDispatcherService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.Jobs.Resources;
using System.Threading;
using Fantasy.ServiceModel;

namespace Fantasy.Jobs.Management
{
    public class StandaloneJobDispatcherService : AbstractService, IJobDispatcher
    {
        public override void InitializeService()
        {
            this._queue = (IJobQueue)this.Site.GetService(typeof(IJobQueue));
            this._filters = AddIn.CreateObjects<IJobStartupFilter>("jobService/startupFilters/filter");
            this._waitHandle = new AutoResetEvent(false);

            _startJobThread = ThreadFactory.CreateThread(this.Run);

            foreach (IJobStartupFilter filter in this._filters)
            {
                if (filter is IObjectWithSite)
                {
                    ((IObjectWithSite)filter).Site = this.Site;
                }
            }

            _queue.Added += new EventHandler<JobQueueEventArgs>(QueueChanged);
            _queue.Changed += new EventHandler<JobQueueEventArgs>(QueueChanged);
            _queue.RequestCancel += new EventHandler<JobQueueEventArgs>(QueueRequestCancel);
            _queue.RequestSuspend += new EventHandler<JobQueueEventArgs>(QueueRequestSuspend);
            _queue.RequestUserPause += new EventHandler<JobQueueEventArgs>(QueueRequestUserPause);

            _resourceQueue = this.Site.GetService<IResourceRequestQueue>();
            _resourceManager = this.Site.GetService<IResourceManager>();
            if (_resourceManager != null)
            {
                _resourceManager.Available += new EventHandler(ResourceAvailable);
            }

            this._controller = this.Site.GetRequiredService<IJobController>();

            base.InitializeService();
        }

        void ResourceAvailable(object sender, EventArgs e)
        {
            this._waitHandle.Set();
        }

        public override void UninitializeService()
        {
            this._startJobThread.Abort(); 
            base.UninitializeService();
        }

        private AutoResetEvent _waitHandle;

        private Thread _startJobThread;

        private IJobQueue _queue;
        private IResourceRequestQueue _resourceQueue;
        private IResourceManager _resourceManager;

        private IJobStartupFilter[] _filters;

        private IJobController _controller;

        public void Start()
        {
            _startJobThread.Start(); 
        }

        void QueueRequestUserPause(object sender, JobQueueEventArgs e)
        {
            IJobController controller = (IJobController)this.Site.GetService(typeof(IJobController));
            controller.UserPause(e.Job.Id);
        }

        void QueueRequestSuspend(object sender, JobQueueEventArgs e)
        {
            IJobController controller = (IJobController)this.Site.GetService(typeof(IJobController));
            controller.Suspend(e.Job.Id);
        }

        void QueueRequestCancel(object sender, JobQueueEventArgs e)
        {
            IJobController controller = (IJobController)this.Site.GetService(typeof(IJobController));
            controller.Cancel(e.Job.Id);
        }

        void QueueChanged(object sender, JobQueueEventArgs e)
        {
            this._waitHandle.Set();
        }

        public void TryDispatch()
        {
            _waitHandle.Set();
        }

        public void Run()
        {
            ILogger logger = this.Site.GetService<ILogger>(); 
            while (true)
            {
               
                try
                {
                    
                    while (TryStartAJob()) ;
                }
                catch (ThreadAbortException)
                {
                }
                catch(Exception error)
                {
                    if (logger != null)
                    {
                        logger.LogError(LogCategories.Manager, error, "An error occurs when try start a new job.");  
                    }
                }
                _waitHandle.WaitOne();
            }
        }

        private bool _starting = false;

        private bool TryStartAJob()
        {
            bool rs = false;
            if (!_starting)
            {
                _starting = true;
                try
                {
                    IEnumerable<JobMetaData> jobs = this._queue.Unterminates;
                    foreach (IJobStartupFilter filter in this._filters)
                    {
                        jobs = filter.Filter(jobs);
                    }

                    JobMetaData job = jobs.FirstOrDefault();
                    if (job != null)
                    {
                        if (this._resourceQueue != null)
                        {
                            this._resourceQueue.UnregisterResourceRequest(job.Id);
                        }
                        ILogger logger = this.Site.GetService<ILogger>();
                        if (job.State == JobState.Unstarted)
                        {
                            job.State = JobState.RequestStart;
                            if (logger != null)
                            {
                                logger.LogMessage("Dispatching", "Start Job {0} ({1}).", job.Name, job.Id);
                            }
                          

                            _controller.StartJob(job);
                        }
                        else
                        {
                            job.State = JobState.RequestStart;
                            if (logger != null)
                            {
                                logger.LogMessage("Dispatching", "Resume Job {0} ({1}).", job.Name, job.Id);
                            }
                            _controller.Resume(job);
                        }

                        rs = true;
                    }
                }
                finally
                {
                    _starting = false;
                }
               
            }

            return rs;

        }



    }
}

File: branches/Stheno/Fantasy.Jobs/src/fantasy/jobs/management/StandaloneJobDispatcherService.java
package fantasy.jobs.management;

import java.rmi.*;
import java.util.*;

import fantasy.collections.*;
import fantasy.jobs.*;
import fantasy.jobs.properties.Resources;
import fantasy.jobs.resources.*;
import fantasy.servicemodel.*;
import fantasy.*;

public class StandaloneJobDispatcherService extends AbstractService implements IJobDispatcher, IJobQueueListener
{
	/**
	 * 
	 */
	private static final long serialVersionUID = -9077723331377116255L;

	public StandaloneJobDispatcherService() throws RemoteException {
		super();
		
	}
	
	
	@Override
	public void initializeService() throws Exception
	{
		this._queue = (IJobQueue)this.getSite().getRequiredService(IJobQueue.class);
		this._filters = AddIn.CreateObjects(IJobStartupFilter.class, "jobService/startupFilters/filter");
		this._waitHandle = new Object();

		_startJobThread = ThreadFactory.create(new Runnable(){

			@Override
			public void run() {
				try {
					StandaloneJobDispatcherService.this.Run();
				} catch (Exception e) {
					
					//e.printStackTrace();
				}
				
			}});

		for (IJobStartupFilter filter : this._filters)
		{
			if (filter instanceof IObjectWithSite)
			{
				((IObjectWithSite)filter).setSite(this.getSite());
			}
		}
		
		this._queue.addListener(this);
		
		


		_resourceQueue = this.getSite().getService(IResourceRequestQueue.class);
		_resourceManager = this.getSite().getService(IResourceManager.class);
		if (_resourceManager != null)
		{
            _resourceManager.addListener(new IResourceManagerListener(){
				@Override
				public void available(EventObject e) {
					ResourceAvailable();
					
				}});

		}

		this._controller = this.getSite().getRequiredService(IJobController.class);

		super.initializeService();
	}

	private void ResourceAvailable()
	{
		synchronized(this._waitHandle)
		{
		    this._waitHandle.notifyAll();
		}
	}

	@Override
	public void uninitializeService() throws Exception
	{
		this._startJobThread.interrupt();
		super.uninitializeService();
	}

	private Object _waitHandle;

	private Thread _startJobThread;

	private IJobQueue _queue;
	private IResourceRequestQueue _resourceQueue;
	private IResourceManager _resourceManager;

	private IJobStartupFilter[] _filters;

	private IJobController _controller;

	public final void start()
	{
		_startJobThread.start();
	}

	public  void RequestUserPause(JobMetaData job) throws Exception
	{
		IJobController controller = (IJobController)this.getSite().getRequiredService(IJobController.class);
		controller.UserPause(job.getId());
	}

	public void RequestSuspend(JobMetaData job) throws Exception
	{
		IJobController controller = (IJobController)this.getSite().getRequiredService(IJobController.class);
		controller.Suspend(job.getId());
	}

	public void RequestCancel(JobMetaData job) throws Exception
	{
		IJobController controller = (IJobController)this.getSite().getRequiredService(IJobController.class);
		controller.Cancel(job.getId());
	}

	public void Changed(JobMetaData job)
	{
		synchronized(this._waitHandle)
		{
		    this._waitHandle.notifyAll();
		}
	}
	
	public void Added(JobMetaData job) throws Exception
	{
		synchronized(this._waitHandle)
		{
		    this._waitHandle.notifyAll();
		}
	}

	public final void tryDispatch()
	{
		synchronized(this._waitHandle)
		{
		    this._waitHandle.notifyAll();
		}
	}

	public final void Run() throws Exception
	{
		ILogger logger = this.getSite().getService(ILogger.class);
		while (true)
		{

			try
			{

				while (TryStartAJob())
				{
					if(Thread.interrupted())
					{
						throw new InterruptedException();
					}
				}
				synchronized(this._waitHandle)
				{
				   _waitHandle.wait();
				}
			}
			catch (InterruptedException e)
			{
				throw e;
			}
			catch(Exception error)
			{
				if (logger != null)
				{
					logger.LogError(LogCategories.getManager(), error, "An error occurs when try start a new job.");
				}
			}
			
		}
	}

	private boolean _starting = false;

	private boolean TryStartAJob() throws Exception
	{
		boolean rs = false;
		if (!_starting)
		{
			_starting = true;
			try
			{
				Iterable<JobMetaData> jobs = this._queue.getUnterminates();
				for (IJobStartupFilter filter : this._filters)
				{
					jobs = filter.Filter(jobs);
				}

				JobMetaData job = new Enumerable<JobMetaData>(jobs).firstOrDefault();
				if (job != null)
				{
					if (this._resourceQueue != null)
					{
						this._resourceQueue.unregisterResourceRequest(job.getId());
					}
					ILogger logger = this.getSite().getService(ILogger.class);
					if (job.getState() == JobState.Unstarted)
					{
						job.setState(JobState.RequestStart);
						if (logger != null)
						{
							logger.LogMessage(LogCategories.getDispatching(), Resources.getStartJobMessage(), job.getName(), job.getId());
						}


						_controller.StartJob(job);
					}
					else
					{
						job.setState(JobState.RequestStart);
						if (logger != null)
						{
							logger.LogMessage(LogCategories.getDispatching(), Resources.getResumeJobMessage(), job.getName(), job.getId());
						}
						_controller.Resume(job);
					}

					rs = true;
				}
			}
			finally
			{
				_starting = false;
			}

		}

		return rs;

	}



}